### PR TITLE
Replace guava fork with caffeine

### PR DESCRIPTION
**Pull request by luxe**
Use upstream library caffeine to replace `ShardActionCache` and `ShardInstance` guava usage.
See discussion: https://github.com/bazelbuild/bazel-buildfarm/issues/718

### CODE REVIEW

File: defs.bzl [hunk 1]
```diff
@@ -77,6 +77,8 @@ def buildfarm_init(name = "buildfarm"):
                         "com.googlecode.json-simple:json-simple:1.1.1",
                         "com.google.code.gson:gson:2.8.6",
                         "com.jayway.jsonpath:json-path:2.4.0",
+                        "com.github.ben-manes.caffeine:caffeine:2.9.0",
+                        "net.javacrumbs.future-converter:future-converter-java8-guava:1.2.0",
```
Comment (werkt) [inline defs.bzl:81]: alpha order this below junit

File: src/main/java/build/buildfarm/instance/shard/ShardActionCache.java [hunk 2]
```diff
@@ -17,47 +17,53 @@
...
+
+          @Override
+          public CompletableFuture<ActionResult> asyncLoad(ActionKey actionKey, Executor executor) {
+            return toCompletableFuture(
+                catching(
+                    service.submit(() -> backplane.getActionResult(actionKey)),
+                    IOException.class,
+                    e -> {
+                      throw Status.fromThrowable(e).asRuntimeException();
+                    },
```
Comment (werkt) [inline src/main/java/build/buildfarm/instance/shard/ShardActionCache.java:55]: We should investigate if this grpc-awareness is actually necessary - it really belongs at the service level to transform if used in a response. Also, this is a little cleaner as a non-throw catchingAsync.

File: src/main/java/build/buildfarm/instance/shard/ShardActionCache.java [hunk 3]
```diff
@@ -76,11 +82,11 @@ public void put(ActionKey actionKey, ActionResult actionResult) {
 
   @Override
   public void invalidate(ActionKey actionKey) {
-    actionResultCache.invalidate(actionKey);
+    actionResultCache.synchronous().invalidate(actionKey);
   }
 
   @Override
   public void readThrough(ActionKey actionKey, ActionResult actionResult) {
```
Comment (werkt) [inline src/main/java/build/buildfarm/instance/shard/ShardActionCache.java:89]: I still question this name, maybe I'm off base here.

File: src/main/java/build/buildfarm/instance/shard/ShardActionCache.java [hunk 4]
```diff
@@ -17,47 +17,53 @@
...
+        };
+
+    actionResultCache = Caffeine.newBuilder().maximumSize(maxLocalCacheSize).buildAsync(loader);
   }
 
   @Override
   public ListenableFuture<ActionResult> get(ActionKey actionKey) {
     return catching(
-        checkNotNull(actionResultCache.get(actionKey)),
+        checkNotNull(toListenableFuture(actionResultCache.get(actionKey))),
```
Comment (werkt) [inline src/main/java/build/buildfarm/instance/shard/ShardActionCache.java:66]: I believe the behavior of caffeine is to throw an NPE on anything that would return a null, we might not need the check.

File: src/main/java/build/buildfarm/instance/shard/BUILD [hunk 5]
```diff
@@ -34,5 +34,6 @@ java_library(
         "@remote_apis//:build_bazel_remote_execution_v2_remote_execution_java_grpc",
         "@remote_apis//:build_bazel_remote_execution_v2_remote_execution_java_proto",
         "@maven//:org_redisson_redisson",
+        "@maven//:net_javacrumbs_future_converter_future_converter_java8_guava",
```
Comment (werkt) [inline src/main/java/build/buildfarm/instance/shard/BUILD:37]: Alpha order deps here and above.

File: src/main/java/build/buildfarm/instance/shard/ShardInstance.java [hunk 6]
```diff
@@ -1195,21 +1197,22 @@ boolean next() {
...
-              public ListenableFuture<Directory> call() {
-                logger.log(
-                    Level.FINE,
-                    format(
-                        "transformQueuedOperation(%s): fetching directory %s",
-                        reason, DigestUtil.toString(directoryBlobDigest)));
-                return fetcher.get();
-              }
-            }),
+        toListenableFuture(directoryCache.get(directoryBlobDigest, getCallback)),
```
Comment (werkt) [inline src/main/java/build/buildfarm/instance/shard/ShardInstance.java:1215]: I don't think this catching is necessary at this point. As indicated by @ben-manes, I believe that a missing entry from caffeine will return a null. Please verify with tests and ensure we've got coverage.

File: src/main/java/build/buildfarm/instance/shard/ShardInstance.java [hunk 7]
```diff
@@ -1239,15 +1242,17 @@ boolean next() {
       Digest commandBlobDigest, Executor executor, RequestMetadata requestMetadata) {
     Supplier<ListenableFuture<Command>> fetcher =
         () -> notFoundNull(expect(commandBlobDigest, Command.parser(), executor, requestMetadata));
+
+    BiFunction<Digest, Executor, CompletableFuture<Command>> getCallback =
+        new BiFunction<Digest, Executor, CompletableFuture<Command>>() {
+          @Override
+          public CompletableFuture<Command> apply(Digest digest, Executor executor) {
+            return toCompletableFuture(fetcher.get());
+          }
```
Comment (werkt) [inline src/main/java/build/buildfarm/instance/shard/ShardInstance.java:1251]: Should we not be using the supplied executor in the fetcher's call to expect at this point?

File: src/main/java/build/buildfarm/instance/shard/ShardInstance.java [hunk 8]
```diff
@@ -193,15 +196,14 @@
...
-      recentCacheServedExecutions =
-          com.google.common.cache.CacheBuilder.newBuilder().maximumSize(64 * 1024).build();
+  private final AsyncCache<Digest, Directory> directoryCache =
+      Caffeine.newBuilder().newBuilder().maximumSize(64 * 1024).buildAsync();
+  private final AsyncCache<Digest, Command> commandCache =
+      Caffeine.newBuilder().newBuilder().maximumSize(64 * 1024).buildAsync();
+  private final AsyncCache<Digest, Action> actionCache =
+      Caffeine.newBuilder().newBuilder().maximumSize(64 * 1024).buildAsync();
+  private final AsyncCache<RequestMetadata, Boolean> recentCacheServedExecutions =
+      Caffeine.newBuilder().newBuilder().maximumSize(64 * 1024).buildAsync();
```
Comment (werkt) [inline src/main/java/build/buildfarm/instance/shard/ShardInstance.java:206]: This should not be an async cache - note the use of the explicit guava upstream synchronous cache type. If you want to use caffeine for everything (I suppose I'm fine with that) then at least use the sync versions of their caches.

I also don't know how you got away with this without changing any of the usages, unless they default to synchronous, and we're only using the sync api.

File: src/main/java/build/buildfarm/operations/finder/BUILD [hunk 9]
```diff
@@ -8,7 +8,7 @@ java_library(
     deps = [
         "//src/main/java/build/buildfarm/ac",
         "//src/main/java/build/buildfarm/common",
-        "//src/main/java/build/buildfarm/common/cache",
+        "@maven//:com_github_ben_manes_caffeine_caffeine",
```
Comment (werkt) [inline src/main/java/build/buildfarm/operations/finder/BUILD:11]: alpha order

File: src/main/java/build/buildfarm/worker/shard/BUILD [hunk 10]
```diff
@@ -6,7 +6,7 @@ java_library(
         "//src/main/java/build/buildfarm/backplane",
         "//src/main/java/build/buildfarm/cas",
         "//src/main/java/build/buildfarm/common",
-        "//src/main/java/build/buildfarm/common/cache",
+        "@maven//:com_github_ben_manes_caffeine_caffeine",
```
Comment (werkt) [inline src/main/java/build/buildfarm/worker/shard/BUILD:9]: alpha order

File: src/main/java/build/buildfarm/instance/shard/ShardInstance.java [hunk 11]
```diff
@@ -1904,7 +1911,7 @@ private void deliverCachedActionResult(
       Operation operation,
       RequestMetadata requestMetadata)
       throws Exception {
-    recentCacheServedExecutions.put(requestMetadata, true);
+    recentCacheServedExecutions.put(requestMetadata, CompletableFuture.completedFuture(true));
```
Comment (werkt) [inline src/main/java/build/buildfarm/instance/shard/ShardInstance.java:1914]: unnecessary once you swap recentCacheServedExecutions to be synchronous.

**Review by werkt — CHANGES_REQUESTED** [review]

File: defs.bzl (hunk 1 above)
Comment (luxe) [reply defs.bzl]: fixed

File: src/main/java/build/buildfarm/worker/shard/BUILD (hunk 10 above)
Comment (luxe) [reply src/main/java/build/buildfarm/worker/shard/BUILD]: fixed

File: src/main/java/build/buildfarm/operations/finder/BUILD (hunk 9 above)
Comment (luxe) [reply src/main/java/build/buildfarm/operations/finder/BUILD]: fixed

File: src/main/java/build/buildfarm/instance/shard/BUILD (hunk 5 above)
Comment (luxe) [reply src/main/java/build/buildfarm/instance/shard/BUILD]: fixed

File: src/main/java/build/buildfarm/instance/shard/ShardActionCache.java (hunk 3 above)
Comment (luxe) [reply src/main/java/build/buildfarm/instance/shard/ShardActionCache.java:89]: It seemed okay to me.  we can simplify things more when framework is swapped.

File: src/main/java/build/buildfarm/instance/shard/ShardInstance.java (hunk 8 above)
Comment (luxe) [reply src/main/java/build/buildfarm/instance/shard/ShardInstance.java]: fixed. back to regular Cache

File: src/main/java/build/buildfarm/instance/shard/ShardInstance.java (hunk 11 above)
Comment (luxe) [reply src/main/java/build/buildfarm/instance/shard/ShardInstance.java]: fixed

File: src/main/java/build/buildfarm/instance/shard/ShardInstance.java (hunk 7 above)
Comment (luxe) [reply src/main/java/build/buildfarm/instance/shard/ShardInstance.java]: moved fetcher inside of callback to use executor

File: src/main/java/build/buildfarm/instance/shard/ShardActionCache.java (hunk 4 above)
Comment (luxe) [reply src/main/java/build/buildfarm/instance/shard/ShardActionCache.java]: fixed

File: src/main/java/build/buildfarm/instance/shard/ShardActionCache.java (hunk 2 above)
Comment (luxe) [reply src/main/java/build/buildfarm/instance/shard/ShardActionCache.java:54]: which part makes you say it's grpc-aware?  I'm not sure how to adjust this, as it matches to previous implementation.

File: src/main/java/build/buildfarm/instance/shard/ShardInstance.java (hunk 6 above)
Comment (luxe) [reply src/main/java/build/buildfarm/instance/shard/ShardInstance.java]: fixed

File: src/main/java/build/buildfarm/instance/shard/ShardActionCache.java (hunk 2 above)
Comment (werkt) [reply src/main/java/build/buildfarm/instance/shard/ShardActionCache.java:54]: The grpc awareness is endemic to the Status translation. It's my gripe that anything below the service implementation, and above the stub, knows anything about grpc.

**Review by werkt — APPROVED** [review]
gogogo

**Pull request merged**